pkg/tlsx/openssl/openssl_exec.go

```diff
@@ -0,0 +1,181 @@
+	cmd.Args = args
+	cmd.Stderr = &errbuff
+	cmd.Stdout = &outbuff
+	cmd.Stdin = &inbuff
+	inbuff.WriteString("Q")
+
+	if err := cmd.Start(); err != nil {
+		return outbuff.String(), errbuff.String(), fmt.Errorf("failed to start openssl: %v", err)
+	}
+	if err := cmd.Wait(); err != nil && errbuff.Len() == 0 {
+		return outbuff.String(), errbuff.String(), err
+	}
+	return strings.TrimSpace(outbuff.String()), errbuff.String(), nil
+}
+
+// getCiphers returns openssl ciphers
+func getCiphers() ([]string, error) {
+	ciphers := []string{}
+	res, _, err := execOpenSSL(context.TODO(), []string{"ciphers"})
+	if err != nil {
+		return ciphers, err
+	}
+	res = strings.TrimSpace(res)
+	ciphers = append(ciphers, strings.Split(res, ":")...)
+	return ciphers, nil
+}
+
+// read openssl s_client response
+func readResponse(data string) (*Response, error) {
+	response := &Response{}
+	if !strings.Contains(data, "CONNECTED") {
+		// If connected string is not available it
+		// openssl failed completely and did not recover
+		return nil, fmt.Errorf("openssl response does not contain 'CONNECTED' %v", data)
+	}
+	var err1, err2 error
+	// openssl s_client returns lot of data however most of
+	// it can be obtained from parse Certificate
+	response.AllCerts, err1 = parseCertificates(data)
+	// Parse Session Data
+	response.Session, err2 = readSessionData(data)
+
+	var err error
+	switch {
+	case err1 != nil:
+		err = wraperrors(err, err1)
+		fallthrough
+	case err2 != nil:
+		err = wraperrors(err, err2)
+		fallthrough
+	case response != nil && (response.AllCerts == nil || len(response.AllCerts) == 0):
+		err = wraperrors(err, fmt.Errorf("no certificates found:\n%v", err))
+		fallthrough
+	case response != nil && response.Session == nil:
+		err = wraperrors(err, fmt.Errorf("session is empty:\n%v", err))
+		fallthrough
+	case err != nil:
+		// if any of above case is successful
+		// add openssl response
+		err = wraperrors(err, fmt.Errorf("\n%v", data))
+	}
+	return response, err
+}
+
+// read Session Data from openssl response
+func readSessionData(data string) (*Session, error) {
+	respreader := bufio.NewReader(strings.NewReader(data))
+	inFlight := false
+	osession := &Session{}
+
+readline:
+	line, err := respreader.ReadString('\n')
+	if err != nil && err != io.EOF {
+		return nil, wraperrors(err, ErrNoSession)
+	} else if err == io.EOF {
+		return osession, nil
+	}
+	line = strings.TrimSpace(line)
+	if strings.HasPrefix(line, "SSL-Session") {
+		inFlight = true
+		goto readline
+	}
+	if inFlight {
+		switch {
+		case strings.HasPrefix(line, "Protocol"):
+			osession.Protocol = parseSessionValue(line)
+		case strings.HasPrefix(line, "Cipher"):
+			osession.Cipher = parseSessionValue(line)
+		case strings.HasPrefix(line, "Master-Key"):
+			osession.MasterKey = parseSessionValue(line)
+		}
+		if !strings.HasPrefix(line, "Extended master secret") {
+			// read until end of session data
+			goto readline
+		}
+	} else {
+		goto readline
+	}
+	return osession, nil
+}
```

Comment:
@xstp  , thanks for pointing out originally I planned to include `TLS session ticket:` but decided to leave it out in this version . that is the reason why I used reader because parsing Session Ticket would be lot easier with reader than with for loop

```yaml
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-ECDSA-CHACHA20-POLY1305
    Session-ID: BA0193628CBD808A1DC30EAD37C2A55C3C708C23B34DBEC65D61FEA6CC0EE1F8
    Session-ID-ctx: 
    Master-Key: A240A29E1083D7F1D19CB8B04BF7D4BE3328B08DC7DCEAC4EC94CF0EA251F4CA072ABBD2418DC9712F04EB48A657FC6D
    TLS session ticket:
    0000 - c9 54 10 98 26 d9 b4 ec-6f a5 ff 08 7a 0f f9 4a   .T..&...o...z..J
    0010 - 27 d9 ea 69 75 31 9b 60-94 22 cf 92 da 97 f5 ba   '..iu1.`."......
    0020 - 19 d4 1b 93 04 15 a4 0b-78 5a f2 d1 01 af 3d a8   ........xZ....=.
    0030 - 52 b4 0d 90 9a 36 c7 a0-9f 62 c7 55 d6 4c ec 04   R....6...b.U.L..
    0040 - 55 8f ba f9 73 f2 e1 b3-19 4a c9 c9 b7 68 1c 67   U...s....J...h.g
    0050 - 74 28 b8 85 01 08 0b 04-cf 20 48 e0 9c 5a 60 6b   t(....... H..Z`k
    0060 - 5c 89 fc 3f 57 c8 67 34-40 86 ea 10 44 33 b8 0e   \..?W.g4@...D3..
    0070 - b1 b2 a4 93 13 fe c2 07-96 30 0d 68 f2 66 dc 0f   .........0.h.f..
    0081 - <SPACES/NULS>

    Start Time: 1672667682
    Timeout   : 7200 (sec)
    Verify return code: 18 (self signed certificate)
---
```